### PR TITLE
Filter report learners using Elasticsearch

### PR DESCRIPTION
**Pull request by sfentress**
This PR

* adds some new migrations to add useful data to the Elasticsearch database
* creates a new portal API for querying the data, with access restricted by permission forms
* speeds up the filtering of users for reports by using ES to get the list of learner ids
* adds a dynamic filtering page for researchers, with filter options and info that update via
ajax requests to the new API, and data that is restricted by permission forms

https://www.pivotaltracker.com/story/show/143254463
https://www.pivotaltracker.com/story/show/122142819
https://www.pivotaltracker.com/story/show/143600433
https://www.pivotaltracker.com/story/show/144382517
https://www.pivotaltracker.com/story/show/144382515

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -62,6 +62,9 @@
 = require 'classnames'
 = require 'react-input-autosize'
 = require 'react-select'
+= require 'moment'
+= require 'daypicker'
```
Comment (scytacki) [inline app/assets/javascripts/application.js]: I think you can also pull in the daypicker using the rails-assets tool: https://rails-assets.org  That lets you bring any bower package into rails. We are using this in LARA, and daypicker seems to be available in bower.

File: config/settings.sample.yml [hunk 2]
```diff
@@ -4,6 +4,7 @@ development: &non_production_settings
   :authoring_site_url: //authoring.staging.concord.org
   :codap_url: http://codap.concord.org/releases/latest/
   :log_manager_data_interactive_url: //cc-log-manager.herokuapp.com/di
+  :elastic_search_endpoint_url: http://search-has-portal-prod-xruhhhyiv2fugtujtzbgfq7bem.us-east-1.es.amazonaws.com
```
Comment (sfentress) [inline config/settings.sample.yml]: Sounds good, but just to understand for the future -- is there a reason why some urls should be in environment variables, and others (like the codap url etc above) are in settings?

File: app/assets/javascripts/application.js [hunk 3]
```diff
@@ -62,6 +62,9 @@
 = require 'classnames'
 = require 'react-input-autosize'
 = require 'react-select'
+= require 'moment'
```
Comment (sfentress) [inline app/assets/javascripts/application.js]: :+1:

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (sfentress) [inline app/assets/javascripts/application.js]: :+1:

File: config/settings.sample.yml (hunk 2 above)
Comment (sfentress) [inline config/settings.sample.yml]: :+1:

File: db/schema.rb [hunk 4]
```diff
@@ -11,7 +11,7 @@
 #
 # It's strongly recommended to check this file into your version control system.
 
-ActiveRecord::Schema.define(:version => 20170202190333) do
+ActiveRecord::Schema.define(:version => 20170410201106) do
```
Comment (scytacki) [inline db/schema.rb]: When resolving the conflict you should make sure the schema references the larger number.

File: app/controllers/api/v1/report_learners_es_controller.rb [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+    if !ENV['ELASTICSEARCH_URL']
+      return error("Elasticsearch endpoint url not set")
+    end
+
+    esSearchResult = self.class.queryES(params, current_user)
+    render :json => esSearchResult
+  end
+
+  def self.queryES(options, user)
```
Comment (scytacki) [inline app/controllers/api/v1/report_learners_es_controller.rb]: In our ruby we try to use underscores not camel case to separate words.

File: app/assets/stylesheets/daypicker.css [hunk 6]
```diff
@@ -0,0 +1,133 @@
+/* DayPicker styles */
```
Comment (scytacki) [inline app/assets/stylesheets/daypicker.css]: Is this file necessary now that daypicker is pulled in as a gem?

File: app/controllers/report/learner_controller.rb [hunk 7]
```diff
@@ -81,45 +81,33 @@ def setup
     if params[:commit] =~ /update learners/i
       update_learners
     elsif params[:commit] == @button_texts[:logs_query]
-      redirect_to learner_logs_query_path(request.GET.except(:commit))
+      redirect_to learner_logs_query_path(request.GET.merge({:commit => true}))
```
Comment (scytacki) [inline app/controllers/report/learner_controller.rb]: Can you explain this code with a comment? It seems to be copying all of the request parameters to the redirect. Is this when it renders the json page of the log manager query?

File: app/controllers/api/v1/report_learners_es_controller.rb (hunk 5 above)
Comment (sfentress) [inline app/controllers/api/v1/report_learners_es_controller.rb]: :+1:

File: app/assets/stylesheets/daypicker.css (hunk 6 above)
Comment (sfentress) [inline app/assets/stylesheets/daypicker.css]: Yes, I mentioned that in the commit, but the rails-assets bower package does not seem to bring in the styes.

File: app/controllers/report/learner_controller.rb (hunk 7 above)
Comment (sfentress) [inline app/controllers/report/learner_controller.rb]: I didn't make a comment because I didn't understand the original code well enough -- or rather, I understood what it did fine, but it felt like there must be a simpler way to do it, so I couldn't justify why the original code was the way it was.

Both the `/report/learner` and `/report/learner/logs_query` routes use this same page. Both use the same `setup` function. In the `setup` function, if it sees that it has the `commit= logs_query` param, it will redirect to `/report/learner/logs_query`. In order not to create an infinite redirect loop, it strips the `commit= logs_query` param.

In my code, unlike the original, I wanted to only make a query for the learners when the `commit` param was there -- i.e. when the user had clicked a button. But the way this was originally, when the `commit= logs_query` param was stripped we wouldn't make a query for the learners.

This change keeps some value for the `commit` query, while preventing the infinite redirect that whoever wrote this originally wanted to avoid.

I've now added a simple comment saying that.

File: db/schema.rb (hunk 4 above)
Comment (sfentress) [inline db/schema.rb]: This is now rebased to master.

**Comment by sfentress** [comment]
I think the failing test in this PR is unrelated to the PR. It looks like it passed in the /push build.

**Comment by scytacki** [comment]
@sfentress can you add documentation about how you run this locally? For now you can add a new file to the `doc` folder.  I'd guess the main info will be how to run and configure Elastic Search and LogStash.

I think the code in the PR is good.  I don't think we should merge it until we setup a demo of it running completely in AWS. The documentation on how you Log Stash locally will help figure out the best approach in AWS.

**Comment by sfentress** [comment]
@scytacki :+1:

https://github.com/concord-consortium/rigse/blob/filter-report-learners-in-es/doc/elasticsearch/elasticsearch-readme.md
